src/plugins/gs-plugin-eos.c

```diff
@@ -264,6 +265,155 @@ app_is_renamed (GsApp *app)
+
+	/* if we don't know the current locale than all locales should be
+	 * compatible */
+	if (!plugin_locale)
+		return TRUE;
+
+	if (g_strcmp0 (locale, plugin_locale) == 0)
+		return TRUE;
+
+	lang = locale_to_language (locale);
+
+	/* in case one of the given locales only represents the language
+	 * (e.g. 'es') check if it's compatible with the other locale's language
+	 * component */
+	if ((g_strcmp0 (lang, plugin_locale) == 0) ||
+	    (g_strcmp0 (plugin_lang, locale) == 0))
+		return TRUE;
+
+	return FALSE;
+}
+
+static char *
+get_app_locale_cache_name (const char *name)
+{
+	g_autofree char *locale_cache_name = g_strdup (name);
+	guint name_length = strlen (locale_cache_name);
+	char *locale_split = NULL;
+
+	/* locales can be as long as 5 chars (e.g. pt_PT) so  */
+	if (name_length <= 5)
+		return NULL;
+
+	locale_split = g_strrstr (locale_cache_name + name_length - 5, "_");
+	if (locale_split)
+		*locale_split = '\0';
+
+	return g_strdup_printf ("locale:%s", locale_cache_name);
+}
+
+static gboolean
+gs_plugin_locale_cached_app_is_best_match (GsPlugin *plugin,
+					   const char *locale_app_id)
+{
+	GsApp *cached_app = NULL;
+	cached_app = gs_plugin_cache_lookup (plugin, locale_app_id);
+
+	if (!cached_app)
+		return FALSE;
+
+	return g_str_has_suffix (gs_app_get_flatpak_name (cached_app),
+				 gs_plugin_get_locale (plugin));
+}
+
+static void
+gs_plugin_update_locale_cache_app (GsPlugin *plugin,
+				   const char *locale_app_id,
+				   GsApp *app)
+{
+	GsApp *cached_app = gs_plugin_cache_lookup (plugin, locale_app_id);
+
+	if (cached_app) {
+		/* avoid blacklisting the same app that's already cached */
+		if (cached_app == app)
+			return;
+
+		gs_app_add_category (cached_app, "Blacklisted");
+	}
+
+	gs_plugin_cache_add (plugin, locale_app_id, app);
+}
+
+static gboolean
+gs_plugin_eos_blacklist_kapp_if_needed (GsPlugin *plugin, GsApp *app)
+{
+	const guint endless_prefix_len = strlen (ENDLESS_ID_PREFIX);
+	g_autofree char *locale_cache_name = NULL;
+	g_auto(GStrv) tokens = NULL;
+	const char *last_token = NULL;
+	const char *locale = gs_plugin_get_locale (plugin);
+	guint num_tokens = 0;
+	/* getting the app name, besides skipping the '.desktop' part of the id
+	 * also makes sure we're dealing with a Flatpak app */
+	const char *app_name = gs_app_get_flatpak_name (app);
+
+	if (!app_name || !locale ||
+	    !g_str_has_prefix (app_name, ENDLESS_ID_PREFIX))
+		return FALSE;
+
+	locale_cache_name = get_app_locale_cache_name (app_name);
+
+	if (g_str_has_suffix (app_name, locale))
+		goto update_cache_and_leave;
+
+	/* skip if the cached app is already our best  */
+	if (gs_plugin_locale_cached_app_is_best_match (plugin,
+						       locale_cache_name)) {
+		gs_app_add_category (app, "Blacklisted");
+		return TRUE;
+	}
+
+	tokens = g_strsplit (app_name + endless_prefix_len, ".", 0);
+	num_tokens = g_strv_length (tokens);
+
+	/* we need at least 2 tokens: app-name & locale */
+	if (num_tokens < 2)
+		return FALSE;
```

Comment:
This block that splits the name should go right after the one where the prefix is checked.
We never want to blacklist any app that don't have a locale component.